lib/statsd.rb

```diff
@@ -14,14 +14,77 @@
 #   statsd = Statsd.new('localhost').tap{|sd| sd.namespace = 'account'}
 #   statsd.increment 'activate'
 class Statsd
-  class RubyUdpClient
-    attr_reader :key, :sock
+  class BufferedUDPClient < UDPClient
+    @max_packet_size = 511
+    (class << self; attr_accessor :max_packet_size; end)
 
-    def initialize(address, port, key = nil)
+    def initialize(address, port)
+      super(address, port)
+      @buffer = String.new
+    end
+
+    def flush!
+      return unless @buffer.bytesize > 0
+      sock.write(@buffer)
+      @buffer.clear
+    rescue => boom
+      nil
+    end
+
+    def send(msg)
+      flush! if @buffer.bytesize + msg.bytesize > BufferedUDPClient.max_packet_size
+      @buffer << msg
+      @buffer << "\n".freeze
+      nil
```

Comment:
> But because of the newline, the next send call will actually send BufferedUDPClient.max_packet_size + 1 bytes.

This is safe -- the default `max_packet_size` has plenty of headroom to fit inside the most strict MTUs.

> Also, what happens if @buffer.bytesize >= BufferedUDPClient.max_packet_size? Should we prevent that somehow?

I think you mean what happens if `msg.bytesize` doesn't fit in a packet? The case you've written should never happen. If `msg.bytesize >= max_packet_size`, we will transmit a packet which may or may not be too large... There's nothing we can do besides sending the packet and praying that the network doesn't drop it (since `max_packet_size` is not a hard cap). I think always sending the packet (even if it's potentially too large) is a better option than silently discarding it. I'd rather give the network a chance to drop it!